File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/investigation/missions/views/MissionsCompletedButton.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.investigation.missions.views;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * ObjectiveCompletedButton class
 *
 * @author Tritium Gaming Studios
 */
public class MissionsCompletedButton extends androidx.appcompat.widget.AppCompatImageButton {

    @Nullable
    private MissionsSpinner objectivesSpinner = null;

    @ColorInt
    private final int[] colorState = {Color.LTGRAY, Color.RED}; // unselected, selected
    private boolean isEnabled = false;

    /**
     * ObjectiveCompletedButton parameterized constructor
     *
     * @param context
     */
    public MissionsCompletedButton(@NonNull Context context) {
        super(context);

        setColorFilter(colorState[0]);

        setOnClickListener(v -> toggleStateAndIcon());
    }

    /**
     * ObjectiveCompletedButton parameterized constructor
     *
     * @param context
     * @param attrs
     */
    public MissionsCompletedButton(@NonNull Context context, AttributeSet attrs) {
        super(context, attrs);

        setColorFilter(colorState[0]);

        setOnClickListener(v -> toggleStateAndIcon());
    }

    /**
     * ObjectiveCompletedButton parameterized constructor
     *
     * @param context
     * @param attrs
     * @param defStyle
     */
    public MissionsCompletedButton(@NonNull Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        setColorFilter(colorState[0]);

        setOnClickListener(v -> toggleStateAndIcon());
    }

    /**
     * setColorStates method
     *
     * @param unselected
     * @param selected
     */
    public void setColorStates(@ColorInt int unselected, @ColorInt int selected) {
        colorState[0] = unselected;
        colorState[1] = selected;

        if (isEnabled()) {
            setColorFilter(colorState[1]);
        }
        else {
            setColorFilter(colorState[0]);
        }
    }

    /**
     * getColorStates method
     *
     * @return int array of set colors
     */
    public int[] getColorStates() {
        return colorState;
    }

    /**
     * toggleStateAndIcon method
     */
    public void toggleStateAndIcon() {
        if (isEnabled()) {
            setEnabledState(false);
            setColorFilter(colorState[0]);
        } else {
            setEnabledState(true);
            setColorFilter(colorState[1]);
        }
        objectivesSpinner.setCompleted(isEnabled());
    }

    /**
     * setEnabledState method
     *
     * @param isEnabled
     */
    public void setEnabledState(boolean isEnabled) {
        this.isEnabled = isEnabled;
    }

    /**
     * overrideStateAndIcon method
     *
     * @param state
     */
    public void overrideStateAndIcon(boolean state) {
        if (!state) {
            setEnabledState(false);
            setColorFilter(colorState[0]);
        } else {
            setEnabledState(true);
            setColorFilter(colorState[1]);
        }
        objectivesSpinner.setCompleted(isEnabled());
    }

    /**
     * isEnabled method
     *
     * @return if the Button is enabled
     */
    public boolean isEnabled() {
        return isEnabled;
    }

    /**
     * setSpinner method
     *
     * @param objectivesSpinner
     */
    public void setSpinner(MissionsSpinner objectivesSpinner) {
        this.objectivesSpinner = objectivesSpinner;
    }

}

File: app-view/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/activities/investigation/missions/views/objectivelayout/MissionsCompletedButton.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.missions.views.objectivelayout

import android.content.Context
import android.util.AttributeSet
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.common.views.PETImageButton
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.ObjectivesViewModel
import com.tritiumgaming.phasmophobiaevidencepicker.util.ColorUtils.getColorFromAttribute


class MissionsCompletedButton : PETImageButton {

    var objectivesViewModel: ObjectivesViewModel? = null

    var missionId: Int = 0

    constructor(context: Context) : super(context) { initView() }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { initView() }

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) :
            super(context, attrs, defStyle) { initView() }

    private fun initView() { }

    fun init(objectivesViewModel: ObjectivesViewModel?, missionId: Int) {
        this.objectivesViewModel = objectivesViewModel
        this.missionId = missionId

        setState()

        setOnClickListener{
            toggleState()
            missionCompleteButtonListener?.onClick()
        }
    }

    private fun setState(
        isComplete: Boolean =
            objectivesViewModel?.spinnerCompletionStatus?.get(missionId) == true) {
        if(isComplete)
            setColorFilter(getColorFromAttribute(context, R.attr.selectedColor))
        else {
            clearColorFilter()
            invalidate()
        }
    }

    private fun toggleState() {
        objectivesViewModel?.toggleCompletionStatus(missionId)

        setState()
    }

    var missionCompleteButtonListener: MissionCompleteButtonListener? = null
    abstract class MissionCompleteButtonListener {
        abstract fun onClick()
    }
}
